README
Device-Velleman-K8055-Native version 0.01
=========================================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2013 by Jonathan Stowe

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.4 or,
at your option, any later version of Perl 5 you may have available.




Changes
v0.0.6	Mon  8 Aug 15:14:09 BST 2022
	* Fix build
v0.0.5	Sun  7 Aug 09:44:11 BST 2022
	* Upload to zef repository
v0.0.4		Thu 21 Jan 09:56:36 GMT 2021
	* Switch to Distribution::Builder::MakeFromJSON
v0.0.3      Mon  4 Feb 16:07:14 GMT 2019
    * Spring clean and upload to CPAN
v0.0.2		Tue Jan 24 21:17:11 2017 +0000 
	* Make zef compatible
v0.0.1		Tue May 17 16:11:52 2016 +0100 
	* Initial release



test.pl
#!/usr/bin/perl 

use strict;
use warnings;

use lib qw(lib);

use Device::Velleman::K8055::Native;


my $dev = Device::Velleman::K8055::Native->new(address   => 0);

while(1)
{
for my $i ( 0 .. 255 )
{
    $dev->write_all_digital($i) || die "$!\n";
    print $dev->_bitmask(), "\n";
    sleep 10;
}
}


src/test.c
 
#include <string.h>
#include <stdio.h>
#include <usb.h>
#include <assert.h>
#include <sys/time.h>
#include "k8055.h"

#define STR_BUFF 256
#define false 0
#define true 1

extern int DEBUG;

int ia1 = -1;
int ia2 = -1;
int id8 = -1;
int ipid = 0;

int numread = 1;

int debug = 0;

int dbt1 = -1; // (-1 => not to set)
int dbt2 = -1; // (-1 => not to set)

int resetcnt1 = false;
int resetcnt2 = false;

int delay = 0;

int main (int argc,char *params[]) 
{
	int i,result;
	long d=0;
	long a1=0,a2=0;
	long c1=0, c2=0;
	unsigned long int start,mstart=0,lastcall=0;

	if ( OpenDevice(ipid)<0 ) 
	{
			printf("Could not open the k8055 (port:%d)\nPlease ensure that the device is correctly connected.\n",ipid);
			return (-1);
			
	} 
        else 
        {
            int j;
		for ( j = 0; j <1256; j++ )
		{
		int i;
                int k = 0;
		for ( i = 0; i <8; i++ )
                {
			k = k + ( 1<< i);
			usleep(100000);
			result=WriteAllDigital((long)k);
		}
		for ( i = 8; i >= 0; i-- )
                {
			k = k - ( 1<< i);
			usleep(100000);
			result=WriteAllDigital((long)k);
		}
                k = 255;
		for ( i = -1; i <8; i++ )
                {
			k = k - ( 1<< i);
			usleep(100000);
			result=WriteAllDigital((long)k);
		}
		for ( i = 8; i >= 0; i-- )
                {
			k = k + ( 1<< i);
			usleep(100000);
			result=WriteAllDigital((long)k);
		}
		}	
		CloseDevice();
	}
	return 0;
}


lib/Device/Velleman/K8055.pm
use v6;

=begin pod

=head1 NAME

Device::Velleman::K8055 - interface with Velleman USB Experiment Board.

=head1 SYNOPSIS

=begin code

use Device::Velleman::K8055;

my $device = Device::Velleman::K8055.new(address => 0);

# Blink alternate LEDs
react {
    whenever Supply.interval(0.5) -> $i {
        if $i %% 2 {
            $device.set-all-digital(0b10101010);
        }
        else {
            $device.set-all-digital(0b01010101);
        }
    }
    whenever signal(SIGINT) {
        $device.close(:reset);
        exit;
    }
}

=end code

=head1 DESCRIPTION

The Velleman K8055 is an inexpensive PIC based board that allows you
to control 8 digital and 2 analogue outputs and read five digital and
2 analog inputs via USB.  There are LEDs on the outputs that show the
state of the outputs (which is largely how I've tested this.)

I guess it would be useful for experimenting or prototyping but it's
rather big (about three times as large as a Raspberry Pi) so you may be
rather constrained if you want to use it in a project.

This module has a fairly simple interface - I guess that a higher level
abstraction could be provided but I only made it as an experiment and
am not quite sure what interface would be best yet.

I've used the L<k8055 library by Jakob
Odersky|https://github.com/jodersky/k8055> to do the low-level parts
rather than binding libusb directly, but all the information is there
is someone else wants to do that.


=head1 METHODS

=head2 method new

    method new(Int :$!address where 0 <= * < 4 = 0, Bool :$debug)

The constructor of the class.  This will attempt to open the
device, throwing an exception if it is unable.  The C<address>
parameter can be used if there is more than one board plugged
in (the default is 0, the first board present.) The C<:debug>
parameter will cause the underlying library to output diagnostic
information to STDERR, so you probably want to use it sparingly.

=head2 method close

    method close(Bool :$reset = False)

This closes the device, freeing up any resources.  If the
C<:reset> parameter is provided, the outputs will be set
to 0 switching off the built in LEDs.

=head2 method set-all-digital

    method set-all-digital(Int $bitmask where * < 256) returns Bool

This sets all the digital outputs based on the bitmask (in the range
0 - 255,) where each of the eight bits represents a single digital
output, with the least significant bit being output 1 (nearest the
edge of the board,) and so forth.

=head2 method set-digital

        method set-digital(Int $channel where * < 8, Bool $v) returns Bool

This sets an individual digital channel, numbered 0 - 7 where 0 is
output 1, setting True turns the output on and False off.

=head2 method set-all-analog

        method set-all-analog(AnalogValue $analog0, AnalogValue $analog1) returns Bool

This sets the analog outputs to a voltage in the range 0-5 volts based on the supplied
values in the range 0-255, I'm not sure how accurate it is.

=head2 method set-analog

        method set-analog(Int $channel where 2 > * => 0, AnalogValue $value) returns Bool

This sets the specified analog channel (0 or 1) to the specified value as described above.

=head2 method reset-counter

        method reset-counter(Int $counter where 2 > * => 0) returns Bool

The board provides two counters on the first two digital inputs.  This resets the specifed
counter to 0.

=head2 method set-debounce-time

        method set-debounce-time(Int $counter where 2 > * => 0, Int $debounce where * < 7450) returns Bool {

This sets the timer for the built in "debounce" of the counters on digital inputs 1 or 2, the debounce time
is a value 0..7450 (in milliseconds.)

=head2 method get-all-input

        method get-all-input(Bool :$quick = False)

This returns a list of the five input values which are:

=item digitalbitmask - a five bit integer indicating the state of the five digital inputs.

=item analog0 - first analog input

=item analog1 - second analog input

=item counter0 - first counter

=item counter1 - second counter


If the C<quick> parameter is supplied, the values may be those buffered and not
reflect the actual state of the inputs.


=head2 method get-all-output

        method get-all-output()

The board itself doesn't provide a mechanism to get the output values, and this is
emulated from the internal cache used by the library.  A list of five values is
returned:

=item digitalBitmask - bitmask value of digital outputs (there are 8 digital outputs)

=item analog0 - value of first analog output

=item analog1 - value of second analog output

=item debounce0 - value of first counter's debounce time [ms]

=item debounce1 - value of second counter's debounce time [ms]

=end pod

use NativeCall;

class Device::Velleman::K8055 {

    constant LIB = %?RESOURCES<libraries/k8055>.Str;

    has Int $.address;

    enum Error (
                   SUCCESS => 0,
                   ERROR => -1,
                   INIT_LIBUSB => -2,
                   NO_DEVICES => -3,
                   NO_K8055 => -4,
                   ACCESS => -6,
                   OPEN => -7,
                   CLOSED => -8,
                   WRITE => -9,
                   READ => -10,
                   INDEX => -11,
                   MEM => -12
                );

    my class Device is repr('CPointer') {


        sub k8055_close_device(Device $device) is native(LIB)  { * }

        method close() {
            k8055_close_device(self);
        }

        sub k8055_set_all_digital(Device $device, int32 $bitmask --> int32 ) is native(LIB) { * }

        method set-all-digital(Int $bitmask where * < 256 --> Bool ) {
            my $rc = k8055_set_all_digital(self, $bitmask);
            $rc == SUCCESS;
        }

        sub k8055_set_digital(Device $device, int32 $channel, bool  $value --> int32 ) is native(LIB) { * }

        method set-digital(Int $channel where * < 8, Bool $v --> Bool ) {
            my $rc = k8055_set_digital(self, $channel, $v.value);
            $rc == SUCCESS;
        }

        sub k8055_set_all_analog(Device $device, int32 $analog0, int32 $analog1 --> int32 ) is native(LIB) { * }

        subset AnalogValue of Int where * < 256;

        method set-all-analog(AnalogValue $analog0, AnalogValue $analog1 --> Bool ) {
            my $rc = k8055_set_all_analog(self, $analog0, $analog1);
            $rc == SUCCESS;
        }

        sub k8055_set_analog(Device $device, int32 $channel, int32 $value --> int32 ) is native(LIB) { * }

        method set-analog(Int $channel where 2 > * >= 0, AnalogValue $value --> Bool ) {
            my $rc = k8055_set_analog(self, $channel, $value);
            $rc == SUCCESS;
        }

        sub k8055_reset_counter(Device $device, int32 $counter --> int32 ) is native(LIB) { * }

        method reset-counter(Int $counter where 2 > * >= 0 --> Bool ) {
            my $rc = k8055_reset_counter(self, $counter);
            $rc == SUCCESS;
        }

        sub k8055_set_debounce_time(Device $device, int32 $counter, int32 $debounce) is native(LIB) { * }

        method set-debounce-time(Int $counter where 2 > * >= 0, Int $debounce where * < 7450 --> Bool ) {
            my $rc = k8055_set_debounce_time(self, $counter, $debounce);
            $rc == SUCCESS;
        }

        sub k8055_get_all_input(Device $device,     int32 $digitalBitmask is rw,
                                                    int32 $analog0 is rw,
                                                    int32 $analog1 is rw,
                                                    int32 $counter0  is rw,
                                                    int32 $counter1 is rw, bool $quick --> int32 ) is native(LIB) { * }

        method get-all-input(Bool :$quick = False) {
            my int32 $digitalBitmask;
            my int32 $analog0;
            my int32 $analog1;
            my int32 $counter0;
            my int32 $counter1;
            my $rc = k8055_get_all_input(self, $digitalBitmask, $analog0, $analog1, $counter0, $counter1, $quick.value);

            if $rc != SUCCESS {
                die "Failed to get inputs ({ Error($rc) })";
            }
            $digitalBitmask, $analog0, $analog1, $counter0, $counter1;
        }

        sub k8055_get_all_output(Device $device,    int32 $digitalBitmask is rw,
                                                    int32 $analog0 is rw,
                                                    int32 $analog1 is rw,
                                                    int32 $debounce0 is rw,
                                                    int32 $debounce1 is rw) is native(LIB)  { * }
        method get-all-output() {
            my int32 $digitalBitmask;
            my int32 $analog0;
            my int32 $analog1;
            my int32 $debounce0;
            my int32 $debounce1;
            k8055_get_all_output(self, $digitalBitmask, $analog0, $analog1, $debounce0, $debounce1);

            $digitalBitmask, $analog0, $analog1, $debounce0, $debounce1;
        }

        method reset() {
            self.set-all-digital(0) && self.set-all-analog(0,0);
        }
    }

    sub k8055_open_device(int32 $port, Pointer $device is rw --> int32 ) is native(LIB) { * }

    method !open-device(Int :$port where { 0 <= $_ < 4 } = 0 --> Device ) {
        my $p = Pointer[Device].new;
        my $rc = k8055_open_device($port, $p);

        if $rc != SUCCESS {
                die "Cannot open device";
        }
        $p.deref;
    }

    method close(Bool :$reset = False) {
        if $reset {
            self.reset;
        }
        $!device.close;
    }

    sub k8055_debug(bool $value) is native(LIB)  { * }

    has Device $!device handles <set-all-digital set-digital set-all-analog set-analog reset-counter set-debounce-time get-all-input get-all-output reset>;

    submethod BUILD(Int :$!address where 0 <= * < 4 = 0, Bool :$debug) {
        $!device = self!open-device(:$!address);
        if $debug {
            k8055_debug(1);
        }
    }

}

# vim: expandtab shiftwidth=4 ft=raku


src/libk8055.c
/* $Id: libk8055.c,v 1.7 2008/08/20 17:00:55 mr_brain Exp $

   This file is part of the libk8055 Library.

   The libk8055 Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The libk8055 Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the
   Free Software Foundation, Inc.,
   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

   http://opensource.org/licenses/

   Copyleft (C) 2005 by Sven Lindberg
     k8055@k8055.mine.nu

   Copyright (C) 2007 by Pjetur G. Hjaltason
       pjetur@pjetur.net
       Commenting, general rearrangement of code, bugfixes,
       python interface with swig and simple k8055 python class


	Input packet format

	+---+---+---+---+---+---+---+---+
	|DIn|Sta|A1 |A2 |   C1  |   C2  |
	+---+---+---+---+---+---+---+---+
	DIn = Digital input in high nibble, except for input 3 in 0x01
	Sta = Status, Board number + 1
	A1  = Analog input 1, 0-255
	A2  = Analog input 2, 0-255
	C1  = Counter 1, 16 bits (lsb)
	C2  = Counter 2, 16 bits (lsb)

	Output packet format

	+---+---+---+---+---+---+---+---+
	|CMD|DIG|An1|An2|Rs1|Rs2|Dbv|Dbv|
	+---+---+---+---+---+---+---+---+
	CMD = Command 
	DIG = Digital output bitmask
	An1 = Analog output 1 value, 0-255
	An2 = Analog output 2 value, 0-255
	Rs1 = Reset counter 1, command 3
	Rs2 = Reset counter 3, command 4
	Dbv = Debounce value for counter 1 and 2, command 1 and 2

	Or split by commands

	Cmd 0, Reset ??
	Cmd 1, Set debounce Counter 1
	+---+---+---+---+---+---+---+---+
	|CMD|   |   |   |   |   |Dbv|   |
	+---+---+---+---+---+---+---+---+
	Cmd 2, Set debounce Counter 2
	+---+---+---+---+---+---+---+---+
	|CMD|   |   |   |   |   |   |Dbv|
	+---+---+---+---+---+---+---+---+
	Cmd 3, Reset counter 1
	+---+---+---+---+---+---+---+---+
	| 3 |   |   |   | 00|   |   |   |
	+---+---+---+---+---+---+---+---+
	Cmd 4, Reset counter 2
	+---+---+---+---+---+---+---+---+
	| 4 |   |   |   |   | 00|   |   |
	+---+---+---+---+---+---+---+---+
	cmd 5, Set analog/digital
	+---+---+---+---+---+---+---+---+
	| 5 |DIG|An1|An2|   |   |   |   |
	+---+---+---+---+---+---+---+---+

**/

#include <string.h>
#include <stdio.h>
#include <usb.h>
#include <assert.h>
#include <math.h>
#include "k8055.h"

#define STR_BUFF 256
#define PACKET_LEN 8

#define K8055_IPID 0x5500
#define VELLEMAN_VENDOR_ID 0x10cf
#define K8055_MAX_DEV 4

#define USB_OUT_EP 0x01	/* USB output endpoint */
#define USB_INP_EP 0x81 /* USB Input endpoint */

#define USB_TIMEOUT 20
#define K8055_ERROR -1

#define DIGITAL_INP_OFFSET 0
#define DIGITAL_OUT_OFFSET 1
#define ANALOG_1_OFFSET 2
#define ANALOG_2_OFFSET 3
#define COUNTER_1_OFFSET 4
#define COUNTER_2_OFFSET 6

#define CMD_RESET 0x00
#define CMD_SET_DEBOUNCE_1 0x01
#define CMD_SET_DEBOUNCE_2 0x01
#define CMD_RESET_COUNTER_1 0x03
#define CMD_RESET_COUNTER_2 0x04
#define CMD_SET_ANALOG_DIGITAL 0x05

/* set debug to 0 to not print excess info */
int DEBUG = 0;

/* variables for usb */
static struct usb_bus *bus, *busses;
static struct usb_device *dev;

/* globals for datatransfer */
struct k8055_dev {
    unsigned char data_in[PACKET_LEN+1];
    unsigned char data_out[PACKET_LEN+1];
    struct usb_dev_handle *device_handle;
    int DevNo;
};

static struct k8055_dev k8055d[K8055_MAX_DEV];
static struct k8055_dev *CurrDev;

/* Keep these globals for now */
unsigned char *data_in, *data_out;

/* char* device_id[]; */

/* Initialize the usb library - only once */
static void init_usb(void)
{
    static int Done = 0;	/* Only need to do this once */
    if (!Done) 
    {
        usb_init();
        usb_find_busses();
        usb_find_devices();
        busses = usb_get_busses();
        Done = 1;
    }
}
/* Actual read of data from the device endpoint, retry 3 times if not responding ok */
static int ReadK8055Data(void)
{
    int read_status = 0, i = 0;

    if (CurrDev->DevNo == 0) return K8055_ERROR;

    for(i=0; i < 3; i++)
        {
        read_status = usb_interrupt_read(CurrDev->device_handle, USB_INP_EP, (char *)CurrDev->data_in, PACKET_LEN, USB_TIMEOUT);
        if ((read_status == PACKET_LEN) && (CurrDev->data_in[1] == CurrDev->DevNo )) return 0;
        if (DEBUG)
            fprintf(stderr, "Read retry\n");
        }
    return K8055_ERROR;
}

/* Actual write of data to the device endpont, retry 3 times if not reponding correctly */
static int WriteK8055Data(unsigned char cmd)
{
    int write_status = 0, i = 0;

    if (CurrDev->DevNo == 0) return K8055_ERROR;

    CurrDev->data_out[0] = cmd;
    for(i=0; i < 3; i++)
        {
        write_status = usb_interrupt_write(CurrDev->device_handle, USB_OUT_EP, (char *)CurrDev->data_out, PACKET_LEN, USB_TIMEOUT);
        if (write_status == PACKET_LEN) return 0;
        if (DEBUG)
            fprintf(stderr, "Write retry\n");
        }
    return K8055_ERROR;
}

/* If device is owned by some kernel driver, try to disconnect it and claim the device*/
static int takeover_device(usb_dev_handle * udev, int interface)
{
    char driver_name[STR_BUFF];
    int ret = K8055_ERROR;

    memset(driver_name, 0, STR_BUFF);

    assert(udev != NULL);
    ret = usb_get_driver_np(udev, interface, driver_name, sizeof(driver_name));
    if (ret == 0)
    {
        if (DEBUG)
            fprintf(stderr, "Got driver name: %s\n", driver_name);
        if (0 > usb_detach_kernel_driver_np(udev, interface))
        {
            if (DEBUG)
                fprintf(stderr, "Disconnect OS driver: %s\n", usb_strerror());
        }
        else if (DEBUG)
            fprintf(stderr, "Disconnected OS driver: %s\n", usb_strerror());
    }
    else if (DEBUG)
        fprintf(stderr, "Get driver name: - %s\n", usb_strerror());

    /* claim interface */
    if (usb_claim_interface(udev, interface) < 0)
    {
        if (DEBUG)
            fprintf(stderr, "Claim interface error: %s\n", usb_strerror());
        return K8055_ERROR;
    }
    else
        usb_set_altinterface(udev, interface);
    usb_set_configuration(udev, 1);

    if (DEBUG)
        {
        fprintf(stderr, "Found interface %d\n", interface);
        fprintf(stderr, "Took over the device\n");
        }

    return 0;
}

/* Open device - scan through usb busses looking for the right device,
   claim it and then open the device
*/
int OpenDevice(long BoardAddress)
{

    int ipid;

    /* init USB and find all of the devices on all busses */
    init_usb();

    /* ID of the welleman board is 5500h + address config */
    if (BoardAddress >= 0 && BoardAddress < K8055_MAX_DEV)
        ipid = K8055_IPID + (int)BoardAddress;
    else
        return K8055_ERROR;              /* throw error instead of being nice */

    /* start looping through the devices to find the correct one */
    for (bus = busses; bus; bus = bus->next)
    {
        for (dev = bus->devices; dev; dev = dev->next)
        {
            if ((dev->descriptor.idVendor == VELLEMAN_VENDOR_ID) &&
                (dev->descriptor.idProduct == ipid))
            {
                CurrDev = &k8055d[BoardAddress];
                CurrDev->device_handle = usb_open(dev);
                if (DEBUG)
                    fprintf(stderr,
                            "Velleman Device Found @ Address %s Vendor 0x0%x Product ID 0x0%x\n",
                            dev->filename, dev->descriptor.idVendor,
                            dev->descriptor.idProduct);
                if (takeover_device(CurrDev->device_handle, 0) < 0)
                {
                    if (DEBUG)
                        fprintf(stderr,
                                "Can not take over the device from the OS driver\n");
                    usb_close(CurrDev->device_handle);   /* close usb if we fail */
                    return K8055_ERROR;  /* throw K8055_ERROR to show that OpenDevice failed */
                }
                else
                {
                    CurrDev->DevNo = BoardAddress + 1; /* Mark as open and valid */
                    SetCurrentDevice(BoardAddress);
                    memset(CurrDev->data_out,0,PACKET_LEN);	/* Write cmd 0, read data */
                    WriteK8055Data(CMD_RESET);
                    if (ReadK8055Data() == 0)
                       return BoardAddress;		/* This function should return board address */
                    else
                       return K8055_ERROR;
                }
            }
        }
    }
    if (DEBUG)
        fprintf(stderr, "Could not find Velleman k8055 with address %d\n",
                (int)BoardAddress);
    return K8055_ERROR;
}

/* Close the Current device */
int CloseDevice()
{
     int rc;

     if (CurrDev->DevNo == 0)
     {
         if (DEBUG)
             fprintf(stderr, "Current device is not open\n" );
         return 0;
     }
     rc = usb_close(CurrDev->device_handle);
     if (rc >= 0)
     {
         CurrDev->DevNo = 0;  /* Not active nay more */
         CurrDev->device_handle = NULL;
     }
     return rc;
}

/* New function in version 2 of Velleman DLL, should return deviceno if OK */
long SetCurrentDevice(long deviceno)
{
    if (deviceno >= 0 && deviceno < K8055_MAX_DEV)
    {
        if (k8055d[deviceno].DevNo != 0)
        {
            CurrDev  = &k8055d[deviceno];
            data_in  = CurrDev->data_in;
            data_out = CurrDev->data_out;
            return deviceno;
        }
    }
    return K8055_ERROR;

}

/* New function in version 2 of Velleman DLL, should return devices-found bitmask or 0*/
long SearchDevices(void)
{
    int retval = 0;
    init_usb();
    /* start looping through the devices to find the correct one */
    for (bus = busses; bus; bus = bus->next)
    {
        for (dev = bus->devices; dev; dev = dev->next)
        {
            if (dev->descriptor.idVendor == VELLEMAN_VENDOR_ID) { 
               if(dev->descriptor.idProduct == K8055_IPID + 0) retval |= 0x01;
               if(dev->descriptor.idProduct == K8055_IPID + 1) retval |= 0x02;
               if(dev->descriptor.idProduct == K8055_IPID + 2) retval |= 0x04;
               if(dev->descriptor.idProduct == K8055_IPID + 3) retval |= 0x08;
            /* else some other kind of Velleman board */
            }
        }
    }
    return retval; 
}

long ReadAnalogChannel(long Channel)
{
    if (Channel == 1 || Channel == 2)
    {
        if ( ReadK8055Data() == 0)
        {
            if (Channel == 2)
                return CurrDev->data_in[ANALOG_2_OFFSET];
            else
                return CurrDev->data_in[ANALOG_1_OFFSET];
        }
        else
            return K8055_ERROR;
    }
    else
        return K8055_ERROR;
}

int ReadAllAnalog(long *data1, long *data2)
{
    if (ReadK8055Data() == 0)
    {
        *data1 = CurrDev->data_in[ANALOG_1_OFFSET];
        *data2 = CurrDev->data_in[ANALOG_2_OFFSET];
        return 0;
    }
    else
        return K8055_ERROR;
}

int OutputAnalogChannel(long Channel, long data)
{
    if (Channel == 1 || Channel == 2)
    {
        if (Channel == 2)
            CurrDev->data_out[ANALOG_2_OFFSET] = (unsigned char)data;
        else
            CurrDev->data_out[ANALOG_1_OFFSET] = (unsigned char)data;

        return WriteK8055Data(CMD_SET_ANALOG_DIGITAL);
    }
    else
        return K8055_ERROR;
}

int OutputAllAnalog(long data1, long data2)
{
    CurrDev->data_out[2] = (unsigned char)data1;
    CurrDev->data_out[3] = (unsigned char)data2;

    return WriteK8055Data(CMD_SET_ANALOG_DIGITAL);
}

int ClearAllAnalog()
{
    return OutputAllAnalog(0, 0);
}

int ClearAnalogChannel(long Channel)
{
    if (Channel == 1 || Channel == 2)
    {
        if (Channel == 2)
            return OutputAnalogChannel(2, 0);
        else
            return OutputAnalogChannel(1, 0);
    }
    else
        return K8055_ERROR;
}

int SetAnalogChannel(long Channel)
{
    if (Channel == 1 || Channel == 2)
    {
        if (Channel == 2)
            return OutputAnalogChannel(2, 0xff);
        else
            return OutputAnalogChannel(1, 0xff);
    }
    else
        return K8055_ERROR;

}

int SetAllAnalog()
{
    return OutputAllAnalog(0xff, 0xff);
}

int WriteAllDigital(long data)
{
    CurrDev->data_out[1] = (unsigned char)data;
    return WriteK8055Data(CMD_SET_ANALOG_DIGITAL);
}

int ClearDigitalChannel(long Channel)
{
    unsigned char data;

    if (Channel > 0 && Channel < 9)
    {
	data = CurrDev->data_out[1] & ~(1 << (Channel-1));
        return WriteAllDigital(data);
    }
    else
        return K8055_ERROR;
}

int ClearAllDigital()
{
    return WriteAllDigital(0x00);
}

int SetDigitalChannel(long Channel)
{
    unsigned char data;

    if (Channel > 0 && Channel < 9)
    {
        data = CurrDev->data_out[1] | (1 << (Channel-1));
        return WriteAllDigital(data);
    }
    else
        return K8055_ERROR;
}

int SetAllDigital()
{
    return WriteAllDigital(0xff);
}

int ReadDigitalChannel(long Channel)
{
    int rval;
    if (Channel > 0 && Channel < 9)
    {
        if ((rval = ReadAllDigital()) == K8055_ERROR) return K8055_ERROR;
        return ((rval & (1 << (Channel-1))) > 0);
    }
    else
        return K8055_ERROR;
}

long ReadAllDigital()
{
    int return_data = 0;

    if (ReadK8055Data() == 0)
    {
        return_data = (
                ((CurrDev->data_in[0] >> 4) & 0x03) |  /* Input 1 and 2 */
                ((CurrDev->data_in[0] << 2) & 0x04) |  /* Input 3 */
                ((CurrDev->data_in[0] >> 3) & 0x18) ); /* Input 4 and 5 */
        return return_data;
    }
    else
        return K8055_ERROR;
}

int ReadAllValues(long int *data1, long int * data2, long int * data3, long int * data4, long int * data5)
{
    if (ReadK8055Data() == 0)
    {
        *data1 = (
                ((CurrDev->data_in[0] >> 4) & 0x03) |  /* Input 1 and 2 */
                ((CurrDev->data_in[0] << 2) & 0x04) |  /* Input 3 */
                ((CurrDev->data_in[0] >> 3) & 0x18) ); /* Input 4 and 5 */
        *data2 = CurrDev->data_in[ANALOG_1_OFFSET];
        *data3 = CurrDev->data_in[ANALOG_2_OFFSET];
        *data4 = *((short int *)(&CurrDev->data_in[COUNTER_1_OFFSET]));
        *data5 = *((short int *)(&CurrDev->data_in[COUNTER_2_OFFSET]));
        return 0;
    }
    else
        return K8055_ERROR;
}

int SetAllValues(int DigitalData, int AdData1, int AdData2)
{
    CurrDev->data_out[1] = (unsigned char)DigitalData;
    CurrDev->data_out[2] = (unsigned char)AdData1;
    CurrDev->data_out[3] = (unsigned char)AdData2;

    return WriteK8055Data(CMD_SET_ANALOG_DIGITAL);
}

int ResetCounter(long CounterNo)
{
    if (CounterNo == 1 || CounterNo == 2)
    {
        CurrDev->data_out[0] = 0x02 + (unsigned char)CounterNo;  /* counter selection */
        CurrDev->data_out[3 + CounterNo] = 0x00;
        return WriteK8055Data(CurrDev->data_out[0]);
    }
    else
        return K8055_ERROR;
}

long ReadCounter(long CounterNo)
{
    if (CounterNo == 1 || CounterNo == 2)
    {
        if (ReadK8055Data() == 0)
        {
            if (CounterNo == 2)
                return *((short int *)(&CurrDev->data_in[COUNTER_2_OFFSET]));
            else
                return *((short int *)(&CurrDev->data_in[COUNTER_1_OFFSET]));
        }
        else
            return K8055_ERROR;
    }
    else
        return K8055_ERROR;
}

int SetCounterDebounceTime(long CounterNo, long DebounceTime)
{
    float value;

    if (CounterNo == 1 || CounterNo == 2)
    {
        CurrDev->data_out[0] = (unsigned char)CounterNo;
        /* the velleman k8055 use a exponetial formula to split up the
           DebounceTime 0-7450 over value 1-255. I've tested every value and
           found that the formula dbt=0,338*value^1,8017 is closest to
           vellemans dll. By testing and measuring times on the other hand I
           found the formula dbt=0,115*x^2 quite near the actual values, a
           little below at really low values and a little above at really
           high values. But the time set with this formula is within +-4% */
        if (DebounceTime > 7450)
            DebounceTime = 7450;
        value = sqrtf(DebounceTime / 0.115);
        if (value > ((int)value + 0.49999999))  /* simple round() function) */
            value += 1;
        CurrDev->data_out[5 + CounterNo] = (unsigned char)value;
        if (DEBUG)
            fprintf(stderr, "Debouncetime%d value for k8055:%d\n",
                    (int)CounterNo, CurrDev->data_out[5 + CounterNo]);
        return WriteK8055Data(CurrDev->data_out[0]);
    }
    else
        return K8055_ERROR;
}

char * Version(void)
{
    return(VERSION);
}
